### ISSUE TITLE
Is it possible to create a mod that adds new buildings or items?

### ISSUE THREAD

**Issue by NeoRider7** [solidDoWant/Planetbase-Framework#12]
Is it possible to create a mod that adds new buildings or items?
For so many years, not a single mod has been added yet that adds new buildings or items.
Why? :-)
Perhaps this is due to the fact that for this we need to involve designers who will make the graphics?
Or is it solely due to problems in programming?

There were many attempts to add new buildings and items, but they remained unfinished.
Example-1: https://steamcommunity.com/workshop/discussions/18446744073709551615/1849197902670717252/?appid=403190
Example-2: https://forums.nexusmods.com/index.php?/topic/9719608-is-this-game-still-alive-for-further-mods/

Is it possible to expect that we will have new mods about buildings and objects in the future? Or is it technically impossible to do so?

**Comment by solidDoWant** [comment]
Yes, currently this is only possible (AFAIK) with my framework, or mods that use my framework. [See here for details](https://github.com/solidDoWant/Planetbase-Framework/wiki/Creating-mods#new-buildings-moduletypes). Alternatively this can be done without any code (similar to challenges) via my XML modloader. In short, a mod creator can build a model in Blender/similar tool, export it as .obj and .mtl files, then bundle it with their mod. The model will then be automagically loaded on startup and can be used for buildings/items/etc.

Here's a couple of proof of concepts I built (artwork/models are placeholders):
* https://steamcommunity.com/sharedfiles/filedetails/?id=1266169042
* https://steamcommunity.com/sharedfiles/filedetails/?id=1266168517

**Comment by NeoRider7** [comment]
Wow, very nice to hear that this is possible!

Clarifying questions:
1) I understand correctly that the main difficulty, why over the years mods with new buildings and objects have not been created, is that there was no designer able to create a model in Blender/similar tool?

2) If we need a person who can work in a Blender, I could try to find one.
But this person will probably ask for a technical task with a description of what he needs to do.
I still don’t quite understand how to write such a technical task for him.

3) Is it possible to see somewhere an example of how .obj and .mtl files are associated with the mod?

4) Regarding the information from that sentence of yours: "**_Alternatively this can be done without any code (similar to challenges) via my XML modloader._**"
- I have a lot of experience with XML and creating challenges. An example of my challenges: [Challenges with ISLAND-shaped maps](https://steamcommunity.com/workshop/filedetails/?id=2032667288).
Is it possible to see examples of such XML code (where there is a mention of .obj and .mtl files)?
I could follow this example to make other XML codes.
I went to the [Planetbase-XMLModloaderMod page](https://github.com/solidDoWant/Planetbase-XMLModloaderMod), but didn't understand how to work with this tool.

Thank you so much for everything you do!

**Comment by solidDoWant** [comment]
To answer your questions:
1. I can only speak to this past 2018, when I first released the feature. Here are the issues with the community in general as I see them, in no particular order. They all affect why "rich" mods have not been developed:
    * Low game and modding popularity. At it's peak (release) the game only had 6000 concurrent players. Nowadays there are 150 - 300 concurrent (https://steamdb.info/app/403190/graphs/). For a game to have an active and vibrant modding community it needs to reach a certain critical mass. If 10% of the players want to mod their game, and 10% of those people want to write mods, that comes out to 15-30 people using mods and 1.5-3 people writing mods (i.e. JPFarias, Tahvohck, and I). There just aren't enough people willing to put in the work to create content.
    * Poor modding community marketing. The PB community is not aware of mods, or think using them is too hard, is not aware of what offerings currently exist, etc. PB modding in general needs more/better marketing to attract players to modding their game, and in turn attracting new mod creators. You're doing a great job of letting people know the current state of the game/modding by answering questions on reddit/steam when they come up, but for the community to thrive we need to attract not just players who are looking to solve a specific problem (i.e. 'how do I make time warp faster?'), but players who are looking to solve a more general problem (i.e. 'how do I increase the replayability of this game?'). I've tried to talk with Martino about this before (a few years ago) but he has been... less than helpful. If he'd be willing to work with us on this it could benefit both him/Madruga Works (he messaged me about building some modding code/tooling a few years ago for dawn of man before it released, this could work for DoM as well) and the players. However I don't think his involvement is likely.
    * Multiple modding frameworks. This goes back to the marketing issue. There are four (arguably five with Melon Loader) modding frameworks that I am aware of: PiMaker's, JPFarias', mine, and Tahvohck's. This leads to confusion and issues like this: https://github.com/solidDoWant/Planetbase-Framework/issues/11. This, in turn, causes people to leave the modding community because it is too difficult/confusing to handle. IMO there needs to be a common platform that supports mods from the other patchers/frameworks. This shouldn't be a [new framework](https://xkcd.com/927/), rather a change to an existing one. Mine should be able to support all JPFarias's mods with a bit of effort. Tahvohck's may be able to as well (I am not familiar with his system so I cannot confirm).
    * Lack of mods. This contributes to the "critical mass" thing mentioned above. Folks won't want to install mods if they don't see any they want, which in turns means that there is nearly 0 chance that they'll write mods, which in turn results in a lack of mods, ... etc.
    
    Circling back to your original question, new buildings/objects have not been created due to non-technical reasons, not technical ones.

2. It's not just that we need a person who can work in Blender/similar modeling program, first we need:
    * A mod idea that is feasible (this is the easiest part). For example, a mod that adds a larger dormitory, as opposed to a mod that turns the game into Kerbel Space Program.
    * Somebody with the knowledge to write the mod, and who's willing to do it (hard to find). That person is probably not me.  I've accomplished what I initially set out to do with PB modding in general and now I don't particularly want to continue given the current state of the community. 
    * Somebody to make art assets, such as 3d models and textures. As you can see from one of my screenshots above, it is pretty trivial to add a new building but making it look good is another thing entirely. I have basically zero 3d modeling or texture making experience but I think it'd be a bit difficult to make buildings look similar to the base game without the help of Madruga Works.

3. Unfortunately I don't have the assets/mod that I was using to test loading when I first implemented it anymore... but it should be similar to the file structure here: https://github.com/solidDoWant/Planetbase-Framework/tree/master/assets. Instead of a "png" or "strings" folder, create a "obj" folder and place the obj and mtl files in there. They will be loaded automagically on startup. If there is interest on a specific example I can build one, I just don't have time today.

4. Here's the short of it:
    1. Install the framework
    2. Install the XMLModloaderMod
    3. Install the XML mod you write at Planetbase\Mods\Xml Modloader
    
    I don't have an example currently (I had one that I was stupid enough to store on hastebin instead of GitHub and it's gone now), but you should be able to read through the [XSD here](https://github.com/solidDoWant/Planetbase-XMLModloaderMod/blob/master/assets/schemas/Version2.xsd) and validate your mod against that.

Let me know if you have further questions.

**Comment by solidDoWant** [comment]
Update in https://github.com/Tahvohck/Planetbase-Mods/issues/4#issuecomment-1035849624

**Comment by NeoRider7** [comment]
![Pics-001-v1 4 0-v01-400x225](https://github.com/solidDoWant/Planetbase-Framework/assets/60851122/d65a7e28-d126-44e2-9491-5c61824182e3)

Hi!  How are you? From the description of the [game update news](https://www.reddit.com/r/Planetbase/comments/1bk1ddl/planetbase_version_140_experimental/) [1.4.0](https://store.steampowered.com/news/app/403190/view/4119177631401549935) about Unity versions: "Planetbase was initially developed in 2015, using Unity 5, this update brings is up to date to **Unity 2022.3**, which is the latest stable. "

In light of this, has anything changed now in your [Framework](https://github.com/solidDoWant/Planetbase-Framework/wiki),  "[Mod Patcher 2](https://www.nexusmods.com/planetbase/mods/35)", add-on "[Planetbase Framework JPFarias compatibility layer](https://www.nexusmods.com/planetbase/mods/57)",  and [mods](https://www.nexusmods.com/planetbase/mods/)?
Maybe now there will be more possibilities to create mods (textures, items, buildings)?
